larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -130,23 +213,117 @@ struct SetBitpackedActivations : public OpRewritePattern<BinaryOp> {
+
+        // Compute new filter values by folding in the per-channel multipliers.
+        std::vector<float> new_filter_values(filter_shape[0] * filter_shape[1] *
+                                             filter_shape[2] * filter_shape[3]);
+        int i = 0;
+        for (float value : filters.getValues<float>()) {
+          float multiplier = filter_per_channel_multipliers
+              [i / (filter_shape[1] * filter_shape[2] * filter_shape[3])];
+          new_filter_values[i++] = value * multiplier;
+        }
+
+        // Create new inputs for the inner bconv op.
+        RankedTensorType filter_type =
+            RankedTensorType::get(filter_shape, rewriter.getF32Type());
+        Value filter_input = rewriter.create<ConstantOp>(
+            inner_bconv_op.filter().getLoc(),
+            DenseElementsAttr::get<float>(filter_type, new_filter_values));
+        RankedTensorType thresholds_type = RankedTensorType::get(
+            {filter_shape[0]}, rewriter.getIntegerType(32));
+        Value thresholds_input = rewriter.create<ConstantOp>(
+            inner_bconv_op.write_bitpacked_threshold().getLoc(),
+            DenseElementsAttr::get<std::int32_t>(thresholds_type, thresholds));
+
+        // We need an empty input with which to overwrite the
+        // `post_activation_multiplier` and `post_activation_bias` (which are no
+        // longer needed, having computed the thresholds).
+        Value empty_input = rewriter.create<ConstantOp>(inner_bconv_op.getLoc(),
+                                                        rewriter.getNoneType(),
+                                                        rewriter.getUnitAttr());
+
         rewriter.replaceOpWithNewOp<TF::LceBconv2dOp>(
-            inner_bconv_op, *maybe_bitpacked_type, inner_bconv_op.getOperands(),
-            inner_bconv_op.getAttrs());
+            inner_bconv_op, *maybe_bitpacked_type, inner_bconv_op.input(),
+            filter_input, empty_input, empty_input, thresholds_input,
+            rewriter.getIntegerAttr(rewriter.getIntegerType(32),
+                                    inner_bconv_op.channels_in()),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(32),
+                                    inner_bconv_op.dilation_height_factor()),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(32),
+                                    inner_bconv_op.dilation_width_factor()),
+            rewriter.getStringAttr(inner_bconv_op.fused_activation_function()),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(32),
+                                    inner_bconv_op.pad_values()),
+            rewriter.getStringAttr(inner_bconv_op.padding()),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(32),
+                                    inner_bconv_op.stride_height()),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(32),
+                                    inner_bconv_op.stride_width()));
```

Comment:
Unfortunately this rather explicit operand/attribute format was necessary to get non-`opt` builds to work properly - the previous method with `ValueRange` lists of operands didn't work. It's a tad uglier, but it works.